fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQuerySetPlan.java

```diff
@@ -152,4 +158,127 @@ default Set<CorrelationIdentifier> tryPushValues(@Nonnull final List<TranslateVa
     default boolean isDynamic() {
         return false;
     }
+
+    static ComparisonKeyFunction.OnValue comparisonKeyValueFunction(@Nonnull final CorrelationIdentifier baseAlias, @Nonnull final Value comparisonKeyValue) {
+        return new ComparisonKeyFunction.OnValue(comparisonKeyValue) {
+            @Override
+            public <M extends Message> Function<QueryResult, List<Object>> apply(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext evaluationContext) {
+                return queryResult -> {
+                    final var nestedContext = evaluationContext.withBinding(baseAlias, queryResult);
+                    return Lists.newArrayList(comparisonKeyValue.eval(store, nestedContext));
+                };
+            }
+        };
+    }
+
+    /**
+     * A comparison key function that extracts a comparison key for binary comparison by some set operations which
+     * also provides a stable plan hash and hash code.
+     */
+    interface ComparisonKeyFunction extends PlanHashable {
```

Comment:
Having introduced this abstraction, I think it may be worth pushing it back to the cursor somehow and making this extend that.